### ISSUE TITLE
Lost inventory

### ISSUE THREAD

**Issue by SlimeDog** [Gnat008/PerWorldInventory#162]
Environment: Spigot 1.11.2 latest, WE/WG, PEX, MV, PWI, WorldResourcepacks all latest. It's been working great forever. Recent issue, though.

The _WorldResourcepacks_ bit is the new factor. I am testing two worlds with different resourcepacks. When I move to any world with the default resources to either world with different resources, I lose all inventory in the "different" world. When I return to any world with default resources, there is inventory in that world. No errors/warnings in log.

If I disable _WorldResourcepacks_, per-world inventories are stable, but I think the problem is for PWI to resolve. As you are probably aware, loading resourcepacks takes many seconds. Best guess is that PWI attempts to load the inventory before the world has loaded, finding...nothing. Assuming there is a method to test for it, please ensure that the world is loaded before grabbing the inventory.

**UPDATE based on later testing (see below):**  It seems that _WorldResourcepacks_ is not the culprit.

**Comment by SlimeDog** [comment]
Looking at the save dates of the inventory files plugins/PerWorldInventory/myUUID/*, it looks like they are not always saved when I change worlds. I posit that the inventories weren't emptied, but reverted to previous inventories, some of which were empty. After recent changes, I rebooted the server. When I returned to the three worlds, the inventories had reverted to the content before the changes. Are the inventory data saved to disk when I change world group, or is it saved only on every (default config) 300 seconds? If the latter, that might (partially) explain the issue.

Bottom line: I have not found a reliable reproduction scenario. Sometimes the inventories are retained, other times not. I would be happy to provide access to my test server, if that would help isolate and resolve this issue.

**Comment by SlimeDog** [comment]
> save-interval: 15

seems to mask the issue. This would not be advisable on a large (or even moderately large) server, but it works fine on my test server, and would probably work on my small whitelisted production server. /top is not reporting heavy load from PWI.

I could try very fast inventory changes and world switches, but this suggests that saving on world group change would resolve the issue. Please let me know what you think.

**Comment by SlimeDog** [comment]
I just lost inventory after logging off, waiting a few seconds (perhaps less than 15), then stopping and re-starting the server.

**Comment by Gnat008** [comment]
That is odd, since when a player disconnects, they should be saved right then and there to the disk. Same when the server shuts down...

**Comment by SlimeDog** [comment]
Nevertheless...

I would be happy to provide access to my test server for debugging.

**Comment by SlimeDog** [comment]
Anything to test?

**Comment by SlimeDog** [comment]
It appears that the inventory is lost on disconnect. Here is the inventory in-game, after changing worlds to elsewhere, then back:

> {"data-format":2,"ender-chest":[],"inventory":{"inventory":[{"index":0,"item":"rO0ABXNyABpvcmcuYnVra2l0LnV0aWwuaW8uV3JhcHBlcvJQR+zxEm8FAgABTAADbWFwdAAPTGph\ndmEvdXRpbC9NYXA7eHBzcgA1Y29t
Lmdvb2dsZS5jb21tb24uY29sbGVjdC5JbW11dGFibGVNYXAk\nU2VyaWFsaXplZEZvcm0AAAAAAAAAAAIAAlsABGtleXN0ABNbTGphdmEvbGFuZy9PYmplY3Q7WwAG\ndmFsdWVzcQB+AAR4cHVyABNbTGphdmEubGFuZy5PYmplY3Q7kM5YnxBzK
WwCAAB4cAAAAAN0AAI9\nPXQABHR5cGV0AAZkYW1hZ2V1cQB+AAYAAAADdAAeb3JnLmJ1a2tpdC5pbnZlbnRvcnkuSXRlbVN0\nYWNrdAAEV09PTHNyAA9qYXZhLmxhbmcuU2hvcnRoTTcTNGDaUgIAAVMABXZhbHVleHIAEGphdmEu\nbGFuZy5O
dW1iZXKGrJUdC5TgiwIAAHhwAA4\u003d\n"},{"index":1,"item":"rO0ABXNyABpvcmcuYnVra2l0LnV0aWwuaW8uV3JhcHBlcvJQR+zxEm8FAgABTAADbWFwdAAPTGph\ndmEvdXRpbC9NYXA7eHBzcgA1Y29tLmdvb2dsZS5jb21tb24uY2
9sbGVjdC5JbW11dGFibGVNYXAk\nU2VyaWFsaXplZEZvcm0AAAAAAAAAAAIAAlsABGtleXN0ABNbTGphdmEvbGFuZy9PYmplY3Q7WwAG\ndmFsdWVzcQB+AAR4cHVyABNbTGphdmEubGFuZy5PYmplY3Q7kM5YnxBzKWwCAAB4cAAAAAJ0AAI9\nP
XQABHR5cGV1cQB+AAYAAAACdAAeb3JnLmJ1a2tpdC5pbnZlbnRvcnkuSXRlbVN0YWNrdAAEV09P\nTA\u003d\u003d\n"},{"index":2,"item":"rO0ABXNyABpvcmcuYnVra2l0LnV0aWwuaW8uV3JhcHBlcvJQR+zxEm8FAgABTAADbWFwdA
APTGph\ndmEvdXRpbC9NYXA7eHBzcgA1Y29tLmdvb2dsZS5jb21tb24uY29sbGVjdC5JbW11dGFibGVNYXAk\nU2VyaWFsaXplZEZvcm0AAAAAAAAAAAIAAlsABGtleXN0ABNbTGphdmEvbGFuZy9PYmplY3Q7WwAG\ndmFsdWVzcQB+AAR4cHVyA
BNbTGphdmEubGFuZy5PYmplY3Q7kM5YnxBzKWwCAAB4cAAAAAN0AAI9\nPXQABHR5cGV0AAZkYW1hZ2V1cQB+AAYAAAADdAAeb3JnLmJ1a2tpdC5pbnZlbnRvcnkuSXRlbVN0\nYWNrdAAEV09PTHNyAA9qYXZhLmxhbmcuU2hvcnRoTTcTNGDaUg
IAAVMABXZhbHVleHIAEGphdmEu\nbGFuZy5OdW1iZXKGrJUdC5TgiwIAAHhwAAs\u003d\n"},{"index":8,"item":"rO0ABXNyABpvcmcuYnVra2l0LnV0aWwuaW8uV3JhcHBlcvJQR+zxEm8FAgABTAADbWFwdAAPTGph\ndmEvdXRpbC9NYX
A7eHBzcgA1Y29tLmdvb2dsZS5jb21tb24uY29sbGVjdC5JbW11dGFibGVNYXAk\nU2VyaWFsaXplZEZvcm0AAAAAAAAAAAIAAlsABGtleXN0ABNbTGphdmEvbGFuZy9PYmplY3Q7WwAG\ndmFsdWVzcQB+AAR4cHVyABNbTGphdmEubGFuZy5PYmp
lY3Q7kM5YnxBzKWwCAAB4cAAAAAN0AAI9\nPXQABHR5cGV0AARtZXRhdXEAfgAGAAAAA3QAHm9yZy5idWtraXQuaW52ZW50b3J5Lkl0ZW1TdGFj\na3QAE1BSSVNNQVJJTkVfQ1JZU1RBTFNzcQB+AABzcQB+AAN1cQB+AAYAAAAHcQB+AAh0AAlt
ZXRh\nLXR5cGV0AAxkaXNwbGF5LW5hbWV0AARsb3JldAAIZW5jaGFudHN0AAlJdGVtRmxhZ3N0AAtVbmJy\nZWFrYWJsZXVxAH4ABgAAAAd0AAhJdGVtTWV0YXQAClVOU1BFQ0lGSUN0ACfCp2JDaGFuZ2Ugd29y\nbGTCp0nCp3TCp2XCp23Cp0r
Cp2/Cp2nCp25zcgA2Y29tLmdvb2dsZS5jb21tb24uY29sbGVjdC5J\nbW11dGFibGVMaXN0JFNlcmlhbGl6ZWRGb3JtAAAAAAAAAAACAAFbAAhlbGVtZW50c3EAfgAEeHB1\ncQB+AAYAAAABdAAZVGVsZXBvcnQgdG8gYW5vdGhlciB3b3JsZHNy
ADdjb20uZ29vZ2xlLmNvbW1v\nbi5jb2xsZWN0LkltbXV0YWJsZUJpTWFwJFNlcmlhbGl6ZWRGb3JtAAAAAAAAAAACAAB4cQB+AAN1\ncQB+AAYAAAABdAALRklSRV9BU1BFQ1R1cQB+AAYAAAABc3IAEWphdmEubGFuZy5JbnRlZ2VyEuKg\npPe
BhzgCAAFJAAV2YWx1ZXhyABBqYXZhLmxhbmcuTnVtYmVyhqyVHQuU4IsCAAB4cAAAAAFzcgAR\namF2YS51dGlsLkhhc2hTZXS6RIWVlri3NAMAAHhwdwwAAAAQP0AAAAAAAAZ0AA5ISURFX1BMQUNF\nRF9PTnQADUhJREVfRU5DSEFOVFN0ABBI
SURFX1VOQlJFQUtBQkxFdAATSElERV9QT1RJT05fRUZG\nRUNUU3QAD0hJREVfQVRUUklCVVRFU3QADUhJREVfREVTVFJPWVN4c3IAEWphdmEubGFuZy5Cb29s\nZWFuzSBygNWc+u4CAAFaAAV2YWx1ZXhwAQ\u003d\u003d\n"}],"armor":[
]},"stats":{"can-fly":true,"display-name":"§4rambeau§r","exhaustion":0.015,"exp":0.96919173,"flying":false,"food":20,"gamemode":"CREATIVE","health":20.0,"level":151,"potion-effects":[],
"saturation":5.0,"fallDistance":0.0,"fireTicks":-20,"maxAir":300,"remainingAir":300}}

and here is the inventory after I disconnect:

> {"data-format":2,"ender-chest":[],"inventory":{"inventory":[],"armor":[]},"stats":{"can-fly":true,"display-name":"§4rambeau§r","exhaustion":0.015,"exp":0.96919173,"flying":false,"food":
20,"gamemode":"CREATIVE","health":20.0,"level":151,"potion-effects":[],"saturation":5.0,"fallDistance":0.0,"fireTicks":-20,"maxAir":300,"remainingAir":300}}

**Comment by Gnat008** [comment]
Now that's really interesting. What could cause that to happen... And it works fine without that one plugin?

**Comment by SlimeDog** [comment]
I will perform a comprehensive test regime, and report back.

**Comment by SlimeDog** [comment]
Test 1: Baseline. Minimum plugin set to retain multiverse:

> Essentials 2.0.1-b467
Multiverse-Core 2.5.0-b725
PermissionsEx 1.23.4
PerWorldInventory 1.9.0
Skript 2.2-dev23
Vault 1.5.6-b49
WorldEdit 6.1.7-b3677
WorldGuard 6.2.1-b1702

World _world_ is in a separate PWI world group.

Test 1.A. Connect/disconnect
- [ ] Connect to world _world_. JSON inventory is empty, in-game inventory is empty.
- [ ] Add three blocks to in-game inventory. JSON inventory remains empty (cached).
- [ ] Disconnect. JSON inventory contains three items, index 0-2.
- [ ] Connect to world _world_. JSON inventory contains three items; in-game inventory is correct.
- [ ] Remove inventory in-game. JSON inventory contains three items (cached).
- [ ] Disconnect. JSON inventory is empty.

Test 1.B. Change worlds
- [ ] Connect to world _world_. JSON inventory is empty, in-game inventory is empty.
- [ ] Add three blocks to in-game inventory. JSON inventory remains empty (cached).
- [ ] Move to world _lobby_. JSON inventory contains three items, index 0-2.
- [ ] Return to world _world_. JSON inventory contains three items; in-game inventory is correct.
- [ ] Disconnect. JSON inventory contains three items, index 0-2.
- [ ] Connect to world _world_. JSON inventory contains three items; in-game inventory is correct.
- [ ] Remove inventory in-game. JSON inventory contains three items (cached).
- [ ] Move to world _lobby_. JSON inventory is empty.
- [ ] Return to world _world_. JSON inventory is empty; in-game inventory is correct.
- [ ] Disconnect. JSON inventory is empty.

All as expected. More to follow.

**Comment by SlimeDog** [comment]
Test 2: Add WorldResourcepacks.

> WorldResourcepacks 1.6.12-b153

Tests performed as in Test 1. All JSON and in-game data correct.

**Comment by SlimeDog** [comment]
Note: With ItemJoin and ChestCommands included, move between worlds using item in inventory slot 9 instead of **/mvtp**.

Test 5: Baseline + ItemJoin + ChestCommands + WorldResourcepacks

>ChestCommands 3.1.4
ItemJoin 3.8
WorldResourcepacks 1.6.12-b153

Protocol performed as indicated. All JSON and in-game data correct.

**Comment by SlimeDog** [comment]
... many more tests ...

**Test 25: All plugins installed** 
except
> -KingKits 5.3.1

**Results:** The anomaly does not occur.

**Test 26: All plugins installed** 
except
> +KingKits 5.3.1
-WorldResourcepacks 1.6.12-b153

**Results:** The anomaly does occur.

**Comment by SlimeDog** [comment]
Resolution: It turned out to be a KingKits configuration issue. Thanks to @KingFaris10 for fast resolution.
